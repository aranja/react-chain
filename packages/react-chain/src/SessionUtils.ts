import { RenderTargetT, SessionT, WrapRenderCallT } from './types'

export function getChainForTarget(
  session: SessionT,
  target?: RenderTargetT,
) {
  switch (target) {
    case 'browser':
      return session.__browserChain
    case 'server':
      return session.__serverChain
    default:
      throw new Error(
        `'${target}' is not a valid render target. ` +
        `A render target can either be 'browser' or 'server'.`
      )
  }
}

export function renderRecursively(wrappers: WrapRenderCallT[], onComplete: Function) {
  let index = 0

  if (wrappers.length === 0) {
    onComplete()
    return
  }

  let didCallRenderers = false

  function wrappedComplete() {
    didCallRenderers = true
    return onComplete.apply(null, arguments)
  }

  function next() {
    const wrap = wrappers[index++]
    const callback = wrappers[index] == null ? wrappedComplete : next
    wrap(callback)
  }

  next()

  if (!didCallRenderers) {
    throw new Error(
      `session.on: some render wrapper didn't call 'render()'.`
    )
  }
}

export function render(session: SessionT, target: RenderTargetT) {
  const chain = getChainForTarget(session, target)
  return (onComplete = () => {}) => renderRecursively(chain, onComplete)
}
