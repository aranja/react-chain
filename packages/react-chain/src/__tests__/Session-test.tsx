import createSession, { InternalSessionT } from '../Session'
import * as React from 'react'
import { render } from '../SessionUtils'

describe('Session', () => {
  let internalSession: InternalSessionT

  beforeEach(() => {
    internalSession = createSession()
  })

  it('should throw if render isn\'t a function', () => {
    expect(() => {
      internalSession.on('browser')
    }).toThrowErrorMatchingSnapshot()
  })

  it('should throw for unknown targets', () => {
    expect(() => {
      internalSession.on()
    }).toThrowErrorMatchingSnapshot()
  })

  it('unfolds correct middleware chain', () => {
    const browserCallOrder: string[] = []
    const serverCallOrder: string[] = []

    internalSession.on('browser', render => {
      browserCallOrder.push('BROWSER_RENDER_1')
      render()
      browserCallOrder.push('BROWSER_RENDER_1')
    })

    internalSession.on('browser', render => {
      browserCallOrder.push('BROWSER_RENDER_2')
      render()
      browserCallOrder.push('BROWSER_RENDER_2')
    })

    internalSession.on('server', render => {
      serverCallOrder.push('SERVER_RENDER_1')
      render()
      serverCallOrder.push('SERVER_RENDER_1')
    })

    internalSession.on('server', render => {
      serverCallOrder.push('SERVER_RENDER_2')
      render()
      serverCallOrder.push('SERVER_RENDER_2')
    })

    render(internalSession, 'browser')(() => {
      browserCallOrder.push('ACTUAL_RENDER')
    })

    render(internalSession, 'server')(() => {
      serverCallOrder.push('ACTUAL_RENDER')
      return ''
    })

    expect(browserCallOrder).toEqual([
      'BROWSER_RENDER_1',
      'BROWSER_RENDER_2',
      'ACTUAL_RENDER',
      'BROWSER_RENDER_2',
      'BROWSER_RENDER_1',
    ])

    expect(serverCallOrder).toEqual([
      'SERVER_RENDER_1',
      'SERVER_RENDER_2',
      'ACTUAL_RENDER',
      'SERVER_RENDER_2',
      'SERVER_RENDER_1',
    ])
  })

  it('should throw if fails to call render', () => {
    internalSession.on('browser', render => {
      render()
    })

    internalSession.on('browser', render => { })

    expect(() => {
      render(internalSession, 'browser')()
    }).toThrow()
  })

  it('can modify the exposed instance', () => {
    const { exposed: session } = internalSession

    if (!session) {
      throw new Error()
    }

    internalSession.on('browser', render => {
      session.someEdit = 'someEdit'
      render()
    })

    internalSession.on('browser', render => {
      session.someEdit += ' anotherEdit'
      render()
    })

    render(internalSession, 'browser')()

    expect(internalSession.exposed).toHaveProperty('someEdit', 'someEdit anotherEdit')
  })

  it('should keep track of current render chain', () => {
    expect(internalSession).toHaveProperty('elementChain')
  })
})
