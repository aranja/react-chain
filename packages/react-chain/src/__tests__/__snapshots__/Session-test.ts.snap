// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Session should throw for unknown targets 1`] = `"session.on: 'unknown' is an invalid render target, it needs to be set to either 'browser' or 'server'."`;
